10-deploying-an-application/README.md

```diff
@@ -0,0 +1,465 @@
+      "Drain connections from ${server.name}",
+      action => 'drain',
+      backend => $instance,
+      server => $server.name,
+    )
+
+    run_task('my_app::deploy', [$server],
+      "Update application for new version",
+    )
+
+    run_task('my_app::lb', $lb_server,
+      action => 'add',
+      backend => $instance,
+      server => $server.name,
+    )
+    notice("Deploy complete on ${server}.")
+  }
+```
+
+Toloop over targets call `get targets` to expand any groups or globs referenced
+in the `$app_servers` parameter then loop over each server with `each`. For
+each server drain connections from the load balancer, deploy the new version of
+application and then adds the server back to the load balancer.  Afterwords log
+a notice message to inform the user that the deploy is complete on that server.
+
+### Perform checks before the deploy
+
+The plan doesn't support multiple load balancers so validate that the
+TargetSpec passed for `$lb_server` resolves to only a single targets and fail
+the plan otherwise. In order to prevent deploys during dangerously high load
+check how many open connections the app has and fail if load is too
+high.
+
+```puppet
+if get_targets($lb_server).length > 1 {
+  fail_plan("${lb_server} did not resolve to a single target")
+}
+
+$conns = run_task('my_app::lb', $lb_server,
+   "Check load before starting deploy",
+   action => 'stats',
+   backend => $instance,
+   server => 'FRONTEND',
+).first['connections']
+if ($conns > 8) {
+  fail_plan("The appplication has too many open connections: ${conns}")
+} else {
+  # Info messages will be displayed when the --verbose flag is used.
+  info("Application has ${conns} open connections.")
+}
+```
+
+Use the `fail_plan` function to stop the plan in both cases. `fail_plan` will
+stop the current plan execution and the execution of any calling plan. It
+accepts a message that will be displayed to the user if the error is not
+caught.
+
+### Check server status before deploy
+
+```puppet
+# Check stats and print a message to the user
+$stats = run_task('my_app::lb', $lb_server,
+  action => 'stats',
+  backend => $instance,
+  server => $server.name,
+  _catch_errors => $force
+).first
+notice("Deploying to ${server.name}, currently ${stats["status"]} with ${stats["connections"]} open connections.")
+```
+
+Before starting to deploy to a server call the `my_app::lb` task with the stats
+action and name of the server that is about to be deployed to and save the
+result the `$stats` variable. Then use those stats to print an informative
+notice about which server is going to be deployed to next.
+
+### Suppress default log messages
+
+Bolt logs every action on each node which results in 5 messages for each node.
+By default these messages are logged at the `notice` level and can make it hard
+to see the more useful `notice` messages the plan logs directly. To make the
+automatic messages log at `info` instead so they won't appear on the console
```

Comment:
s/console/terminal so as not to conflate with puppet console?  Not sure how that could happen, but in the future it seems like it'll become more likely to confuse them.